keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,838 @@
+
+For ResourceClaim, since the field in question is part of an alpha
+feature, we will just update its validation to unconditionally require
+strict validation of all IP values.
+
+```
+<<[UNRESOLVED array ratcheting ]>>
+
+Is this behavior actually what we want, for all cases?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Ratcheting validation for immutable fields
+
+Four of the fields listed above are immutable:
+
+  - `pod.spec.dnsConfig.nameservers[]`
+  - `pod.spec.hostAliases[].ip`
+  - `service.spec.clusterIP`
+  - `service.spec.clusterIPs[]`
+
+It has been suggested that we should allow fixing invalid values in
+such fields. That is, you would be allowed to modify them if:
+
+  - the old value does not pass current validation rules, _and_
+  - the new value is the canonical representation of the old value.
+
+So given `clusterIP: 172.030.099.099`, you would be allowed to modify
+it to `clusterIP: 172.30.99.99`, but not to any other value. (For
+example, you could not modify it to `clusterIP: 172.30.99.099`,
+because while that is less wrong than the original value, it is still
+wrong, and not the canonical representation of that IP.)
+
+This allows people to fix bad values in existing objects, but adds
+complexity to validation, and potentially could cause problems with
+code that was expecting a more literal definition of "immutable".
+
+Also, in the case of the `Pod` fields, it shouldn't be too difficult
+to simply recreate the pods with correct values, and in the case of
+the `Service` fields, it is unlikely that many objects with invalid
+`clusterIP` values exist, since in most cases they are assigned by the
+controller (which never assigns invalid values), and if a user did
+assign an invalid `clusterIP` value by hand, there is already an
+apiserver warning about it.
+
+So maybe there is no real point in doing this.
+
+```
+<<[UNRESOLVED immutable ratcheting ]>>
+
+Do we want to do this? (It is currently implemented, but could be
+removed. It could also potentially be added later even if we didn't
+add it now.)
+
+<<[/UNRESOLVED]>>
+```
+
+### Risks and Mitigations
+
+The new validation should increase the security of Kubernetes by
+making it impossible to have ambiguously-interpretable IPs/CIDRs in
+the future.
+
+The most obvious risk to users is that by tightening validation, we
+might break existing clusters. Especially, the fact that we are
+enforcing tighter validation for new objects of existing types means
+that we might break some users' existing workflows / automation that
+were generating now-invalid values. This will be mitigated by having
+API warnings for a few releases before we flip the switch, and by
+wrapping the new validation in a feature gate that users could disable
+if needed while they update their infrastructure.
+
+The new validation logic (to allow legacy invalid values) will be more
+complicated than the old logic, and thus potentially more likely to
+have bugs.
```

Comment:
Since for the immutable fields you're not planning to change them, there's a risk that these objects will exist in clusters for longer than necessary, potentially leaving cluster vulnerable. The responsibility to cleaning those up will be up to cluster-admins, but we need to make sure this is listed here explicitly.